### PR TITLE
Remap pretrained cls head by class name when fine-tuning 

### PR DESCRIPTION
**Pull request by Y-T-G**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds class-name-based pretrained head remapping for YOLO fine-tuning, improving weight transfer when datasets have different or reordered classes 🎯

### 📊 Key Changes
- Added a new `cls_remap` training option, enabled by default in `default.yaml` ✅
- Introduced class-name-aware remapping for detection classification head weights during model loading 🔁
- Applies remapping only to closed-set class-logit heads, avoiding incorrect changes to YOLOE embedding heads or unrelated tensors 🛡️
- Ensures dataset class names are set on models before loading weights across detection, segmentation, pose, and OBB trainers 🧠
- Added tests covering:
  - Correct remapping when class order differs
  - Safe behavior for YOLOE models
  - Trainer integration for detect, segment, pose, and OBB tasks
- Minor docstring capitalization cleanup in mask-processing tests ✨

### 🎯 Purpose & Impact
- Improves fine-tuning from pretrained models when source and target datasets share some class names but use different class counts or class orders 🚀
- Helps preserve useful pretrained knowledge for matching classes like `cat`, `dog`, or `car`, potentially improving convergence and accuracy 📈
- Reduces risk of loading class weights into the wrong class positions, making transfer learning safer and more reliable 🔒
- Benefits users training YOLO26 and other Ultralytics models on custom datasets with overlapping labels, with no extra action needed unless they disable `cls_remap` ⚙️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution has been received, and an Ultralytics engineer will assist with review.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

No genuine blocking issues found in the diff. The remapping is scoped to detection classification-head tensors, avoids direct intersect loading after remap, and the new config flag is wired into the boolean config keys.

**Comment by Y-T-G** [comment]
With this branch, if we finetune the YOLO26s pretrained on Objects365 on COCO128, we can see immediate mAP headstart:

```
yolo train model=yolo26s-objv1-150.pt data=coco128.yaml epochs=10 device=-1

Remapped 78/80 cls head rows from pretrained weights by class name
Transferred 708/708 items from pretrained weights


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/10      5.09G      1.408      2.049   0.009527        120        640: 100% ━━━━━━━━━━━━ 8/8 1.5it/s 5.5s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 1.3s/it 5.2s
                   all        128        929      0.638      0.523      0.572      0.418

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/10       6.4G      1.364      1.984   0.009352         68        640: 100% ━━━━━━━━━━━━ 8/8 7.3it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.7it/s 0.6s
                   all        128        929      0.623      0.549      0.594      0.443

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/10      6.46G      1.374      2.018   0.009903         96        640: 100% ━━━━━━━━━━━━ 8/8 7.0it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.8it/s 0.6s
                   all        128        929      0.734      0.494      0.607      0.455

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/10      6.51G      1.365      1.775   0.009128        124        640: 100% ━━━━━━━━━━━━ 8/8 7.6it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.6it/s 0.6s
                   all        128        929      0.723      0.515      0.615      0.464

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       5/10      6.56G      1.369      1.727   0.008543         70        640: 100% ━━━━━━━━━━━━ 8/8 6.3it/s 1.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.7it/s 0.6s
                   all        128        929      0.709      0.547      0.629      0.475

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       6/10      6.61G       1.34      1.698   0.009667        145        640: 100% ━━━━━━━━━━━━ 8/8 10.7it/s 0.7s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 5.1it/s 0.8s
                   all        128        929      0.737      0.556      0.648      0.492

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       7/10      6.61G      1.299      1.613   0.008946        111        640: 100% ━━━━━━━━━━━━ 8/8 7.4it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.7it/s 0.6s
                   all        128        929      0.753      0.569      0.658      0.501

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       8/10      6.61G      1.288       1.53   0.008664        160        640: 100% ━━━━━━━━━━━━ 8/8 7.4it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 10.4it/s 0.4s
                   all        128        929      0.756      0.586      0.674      0.512

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       9/10      6.61G      1.256      1.553   0.008553         94        640: 100% ━━━━━━━━━━━━ 8/8 7.5it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.7it/s 0.6s
                   all        128        929      0.773      0.578      0.678      0.522

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      10/10      6.61G      1.246      1.446   0.008267        133        640: 100% ━━━━━━━━━━━━ 8/8 7.7it/s 1.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.3it/s 0.6s
                   all        128        929      0.737      0.604       0.68      0.522

```


But we don't get that benefit with `n` variant since `c3` changes from 64 to 100 depending on `nc` and the weights fail to transfer:

```
yolo train model=yolo26n-objv1-150.pt data=coco128.yaml epochs=10 device=-1

Remapped 78/80 cls head rows from pretrained weights by class name
Transferred 612/708 items from pretrained weights


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/10      2.97G      1.447      5.942    0.00991        120        640: 100% ━━━━━━━━━━━━ 8/8 1.3it/s 5.9s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 1.4s/it 5.4s
                   all        128        929   2.02e-05    0.00305   8.08e-06   2.38e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/10      3.63G      1.387      5.951   0.009274         68        640: 100% ━━━━━━━━━━━━ 8/8 12.0it/s 0.7s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 5.6it/s 0.7s
                   all        128        929   2.46e-05    0.00372   1.55e-05   4.92e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/10      3.66G      1.375      5.963    0.01018         96        640: 100% ━━━━━━━━━━━━ 8/8 8.1it/s 1.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 10.7it/s 0.4s
                   all        128        929   2.71e-05     0.0041   2.45e-05   8.19e-06

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/10      3.68G      1.365      5.901   0.009133        124        640: 100% ━━━━━━━━━━━━ 8/8 6.8it/s 1.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 21.2it/s 0.2s
                   all        128        929   3.08e-05    0.00466   4.44e-05   1.65e-05

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       5/10       3.7G      1.386      5.961   0.008432         70        640: 100% ━━━━━━━━━━━━ 8/8 6.8it/s 1.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 9.8it/s 0.4s
                   all        128        929   3.48e-05    0.00527   7.05e-05   2.97e-05

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       6/10      3.72G      1.332      5.877    0.00917        145        640: 100% ━━━━━━━━━━━━ 8/8 11.6it/s 0.7s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 5.6it/s 0.7s
                   all        128        929   4.14e-05    0.00627   0.000115   5.24e-05

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       7/10      3.72G      1.275      5.942   0.008791        111        640: 100% ━━━━━━━━━━━━ 8/8 7.0it/s 1.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 20.0it/s 0.2s
                   all        128        929   4.36e-05     0.0066   0.000164   8.33e-05

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       8/10      3.72G       1.25      5.871    0.00854        160        640: 100% ━━━━━━━━━━━━ 8/8 15.8it/s 0.5s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 4.2it/s 0.9s
                   all        128        929   4.77e-05    0.00721   0.000259   0.000137

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       9/10      3.72G      1.299       5.85   0.009056         94        640: 100% ━━━━━━━━━━━━ 8/8 10.4it/s 0.8s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 6.7it/s 0.6s
                   all        128        929   5.13e-05    0.00776    0.00036   0.000186

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      10/10      3.72G      1.292      5.793    0.00851        133        640: 100% ━━━━━━━━━━━━ 8/8 7.2it/s 1.1s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 20.5it/s 0.2s
                   all        128        929   5.58e-05    0.00843   0.000405   0.000212
```

The `c3` only changes for `n` variant. For the other variants, it stays fixed because `ch[0]` is larger than 100 for the other variants.

**Comment by UltralyticsAssistant** [comment]
🚀 Merged! Thank you @Y-T-G for adding class-name-based pretrained head remapping, with contributions from @glenn-jocher.

“Genius is one percent inspiration and ninety-nine percent perspiration.” — Thomas Edison

This PR reflects that spirit by turning a subtle transfer-learning challenge into a safer, smarter default for Ultralytics users. The new `cls_remap` behavior helps preserve useful pretrained knowledge when class names overlap but ordering or counts differ, improving fine-tuning reliability across detection, segmentation, pose, and OBB workflows while keeping YOLOE and unrelated heads protected.

A meaningful step forward for users training YOLO26 and other Ultralytics models on custom datasets—thanks for the thoughtful implementation, integration coverage, and tests!
